### ISSUE TITLE
Memory and  semi-hard example mining code

### ISSUE THREAD

**Issue by TheoPis** [tfzhou/ContrastiveSeg#31]
Hi, well done on the nice work. 
I have the following questions regarding the codebase:

1) I believe there is an issue in the pixel memory code. (or perhaps that's intended but not mentioned in the paper ?)
In the case where available samples (K) exceed the memory size for a given class, the code replaces the newest rather than the oldest features in the queue (which is for example how moco handles the memory bank). 
Specifically in enqueue_dequeue in trainer_contrastive.py: 

if ptr + K >= self.memory_size:
                    pixel_queue[lb, -K:, :] = nn.functional.normalize(feat,p=2, dim=1)
                    pixel_queue_ptr[lb] = 0

2) In section 3.3 of the paper it is mentioned that the version of the method that is giving improvemnts on multiple datasets, entails semi-hard negative example mining and as a sampling procedure from the memory bank. I could not find the exact implementation of this version in the repository. 

The only thing I found related to getting negatives from the memory is in loss_contrast_mem.py, method _sample_negative(self, Q) which to the best of my understanding just takes all memory samples per class.

Please correct me if I am missing something here.
If not, are you planning to either add these methods or make trained model weights available?  

3) Finally, It would be helpful to know whether all the miou results in the paper (including the appendix and results that concern ablation on the Cityscapes validation set ) are using multi-scale + flip test time augmentation. 

It would be very helpful if you could address these questions.
Thanks

**Comment by TheoPis** [comment]
Hi again, 
I have noticed the authors are frequently responding to other issues, thus I would appreciate if these questions could also be adressed.
Many thanks

**Comment by nysp78** [comment]
@TheoPis  have you managed to find the implementation of semi-hard sampling?

**Comment by TheoPis** [comment]
@nysp78 to the best of my understanding, no hardness-aware sampling method is in this repo.

**Comment by nysp78** [comment]
@TheoPis  do you have any ideas on where I can find a similar implementation for hardness aware sampling?

**Comment by crisz94** [comment]
A lot of similar questions have been raised through these years, but none of them been responded, lol

**Comment by hitcbw** [comment]
``` python
if ptr + K >= self.memory_size:
      pixel_queue[lb, -K:, :] = nn.functional.normalize(feat,p=2, dim=1)
"""

It doesn't means the update only occurs on the latest samples. When the ptr over than queue size, the ptr will be moved left and more samples are updated. The most early features are updated in the ``else" section.
